subsys/dap/cmsis_dap.c

```diff
@@ -0,0 +1,1062 @@
+			if (rspns_val != SWDP_ACK_OK) {
+				goto end;
+			}
+
+			/* Store data */
+			sys_put_le32(data, rspns_buf);
+			rspns_buf += sizeof(data);
+			rspns_cnt++;
+		}
+	} else {
+		/* Write register block */
+		while (req_cnt--) {
+			/* Load data */
+			data = sys_get_le32(req_buf);
+			req_buf += sizeof(data);
+			/* Write DP/AP register */
+			rspns_val = do_swdp_transfer(ctx, req_val, &data);
+			if (rspns_val != SWDP_ACK_OK) {
+				goto end;
+			}
+
+			rspns_cnt++;
+		}
+		/* Check last write */
+		rspns_val = do_swdp_transfer(ctx, DP_RDBUFF | SWDP_REQUEST_RnW, NULL);
+	}
+
+end:
+	sys_put_le16(rspns_cnt, &response[0]);
+	response[2] = rspns_val;
+
+	LOG_DBG("Received %u, to transmit %u, response count %u",
+		req_buf - request,
+		rspns_buf - response,
+		rspns_cnt * 4);
+
+	return (rspns_buf - response);
+}
+
+/* Delegate Transfer Block command */
+static uint16_t dap_transferblock(struct dap_context *const ctx,
+				  const uint8_t *const request,
+				  uint8_t *const response)
+{
+	uint16_t retval;
+
+	if (!atomic_test_bit(&ctx->state, DAP_STATE_CONNECTED)) {
+		LOG_ERR("DAP device is not connected");
+		/* Clear response count */
+		sys_put_le16(0U, &response[0]);
+		/* Clear DAP response (ACK) value */
+		response[2] = 0U;
+		return 3U;
+	}
+
+	switch (ctx->debug_port) {
+	case DAP_PORT_SWD:
+		retval = dap_swdp_transferblock(ctx, request, response);
+		break;
+	case DAP_PORT_JTAG:
+	default:
+		LOG_ERR("port unsupported");
+		/* Clear response count */
+		sys_put_le16(0U, &response[0]);
+		/* Clear DAP response (ACK) value */
+		response[2] = 0U;
+		retval = 3U;
+	}
+
+	return retval;
+}
+
+/* Process SWD Write ABORT command and prepare response */
+static uint16_t dap_swdp_writeabort(struct dap_context *const ctx,
+				   const uint8_t *const request,
+				   uint8_t *const response)
+{
+	const struct swdp_api *api = ctx->swdp_dev->api;
+	/* Load data (Ignore DAP index in request[0]) */
+	uint32_t data = sys_get_le32(&request[1]);
+
+	/* Write Abort register */
+	api->swdp_transfer(ctx->swdp_dev, DP_ABORT, &data,
+			  ctx->transfer.idle_cycles, NULL);
+
+	response[0] = DAP_OK;
+	return 1U;
+}
+
+/* Delegate DAP Write ABORT command */
+static uint16_t dap_writeabort(struct dap_context *const ctx,
+			       const uint8_t *const request,
+			       uint8_t *const response)
+{
+	uint16_t retval;
+
+	if (!atomic_test_bit(&ctx->state, DAP_STATE_CONNECTED)) {
+		LOG_ERR("DAP device is not connected");
+		response[0] = DAP_ERROR;
+		return 1U;
+	}
+
+	switch (ctx->debug_port) {
+	case DAP_PORT_SWD:
+		retval = dap_swdp_writeabort(ctx, request, response);
+		break;
+	case DAP_PORT_JTAG:
+	default:
+		LOG_ERR("port unsupported");
+		response[0] = DAP_ERROR;
+		retval = 1U;
+	}
+	return retval;
+}
+
+/* Process DAP Vendor command request */
+static uint16_t dap_process_vendor_cmd(struct dap_context *const ctx,
+				       const uint8_t *const request,
+				       uint8_t *const response)
+{
+	response[0] = ID_DAP_INVALID;
+	return 1U;
+}
```

Comment:
Define this as weak so a user can easily override this in their app. That way a user doesn't have to modify code in the zephyr tree.

Reply:
No, there are other ways to achieve it.

Reply:
of course, open to better ways

Reply:
This has not been addressed yet, I know how, but I do not have the time yet.

Reply:
Would like to see this done before this PR is merged.

Reply:
It should be relatively easy to register even a single callback, and that could be done in a follow-up PR. Vendor commands are relatively specialized as well, so why add it in a generic PR?

There are probably other reasons to not block here, maybe creeping into the neighborhood of policy.. 🤷🏼‍♂️

Reply:
Nice point, adding it later is fine. I was trying to push my own agenda because this is a key feature I need :)